db/etcd.go

```diff
@@ -27,499 +27,336 @@ import (
-func (etcddb etcdDB) GetObjects(obj_type string) []objects.ToddObject {
+// This requires an "objType" string to specify the type of object being looked up.
+func (etcddb *etcdDB) GetObjects(objType string) ([]objects.ToddObject, error) {
 
-	var ret_obj []objects.ToddObject
+	retObj := []objects.ToddObject{}
 
-	etcd_loc := fmt.Sprintf("http://%s:%s", etcddb.config.DB.IP, etcddb.config.DB.Port)
-	cfg := client.Config{
-		Endpoints: []string{etcd_loc},
-		Transport: client.DefaultTransport,
-		// set timeout per request to fail fast when the target endpoint is unavailable
-		HeaderTimeoutPerRequest: time.Second,
-	}
-	c, err := client.New(cfg)
-	if err != nil {
-		log.Fatal(err)
-	}
-	kapi := client.NewKeysAPI(c)
-
-	// Construct the path to the key depending on the obj_type param
-	var key_str string
-	if obj_type == "" {
+	// Construct the path to the key depending on the objType param
+	if objType == "" {
 		// TODO(mierdin): support this empty type, and return an entire list of objects regardless of type
-		log.Warn("Object API queried with no type argument -- returning empty slice")
-		var empty []objects.ToddObject
-		return empty
-	} else {
-		//Construct a path to the key based on the provided type
-		key_str = fmt.Sprintf("/todd/objects/%s/", obj_type)
+		return nil, errors.New("Object API queried with no type argument")
 	}
 
-	log.Info("Accessing objects at", key_str)
+	//Construct a path to the key based on the provided type
+	keyStr := fmt.Sprintf("/todd/objects/%s/", objType)
+
+	log.Info("Accessing objects at", keyStr)
 
-	resp, err := kapi.Get(context.Background(), key_str, &client.GetOptions{Recursive: true})
+	resp, err := etcddb.keysAPI.Get(context.Background(), keyStr, &client.GetOptions{Recursive: true})
 	if err != nil {
-		fmt.Println(err)
 		log.Warn("ToDD object store empty when queried")
-	} else {
-
-		log.Debugf("Etcd 'get' is done. Metadata is %q\n", resp)
-
-		// We are expecting that this node is a directory
-		if resp.Node.Dir {
-
-			// Iterate over found objects
-			for i := range resp.Node.Nodes {
-				node := resp.Node.Nodes[i]
-
-				log.Printf("Parsing object %s \n", node.Value)
+		return retObj, nil
+	}
 
-				// Marshal API data into ToddObject
-				var baseobj objects.BaseObject
-				err = json.Unmarshal([]byte(node.Value), &baseobj)
-				if err != nil {
-					log.Error("Failed to retrieve data from server")
-					os.Exit(1)
-				}
+	log.Debugf("Etcd 'get' is done. Metadata is %q\n", resp)
 
-				// Generate a more specific Todd Object based on the JSON data
-				finalobj := baseobj.ParseToddObject([]byte(node.Value))
+	// We are expecting that this node is a directory
+	if !resp.Node.Dir {
+		// We are definitely expecting an ETCd directory, so we should return nothing if this is not the case.
+		return nil, errors.New("Etcd query for objects did not result in a directory as expected")
+	}
 
-				ret_obj = append(ret_obj, finalobj)
+	// Iterate over found objects
+	for _, node := range resp.Node.Nodes {
```

Comment:
Yes, though there is some nuance to where you might use one over the other. This syntax gives you a copy of the element of the slice. So, for example, if you had a slice of structs and want to modify all of them, this wouldn't be appropriate because you'd only modify the copy, not the original in the slice.

https://github.com/golang/go/wiki/Range

Also, making copies may be slower. May be because indexing requires a bounds check, so ymmv depending on how large the element is, how many times you index, and probably a bunch of other stuff I don't know about.(obligatory "premature optimization is the root of all evil" reference here :smile: )